File: sdk/iris_otel/tracer.py
import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

_initialized = False


def _init_tracer_provider() -> None:
    global _initialized
    if _initialized:
        return

    endpoint = os.environ.get("IRIS_OTLP_ENDPOINT", "http://localhost:4318/v1/traces")
    service_name = os.environ.get("IRIS_SERVICE_NAME", "iris-instrumented-app")

    provider = TracerProvider(resource=Resource.create({"service.name": service_name}))
    provider.add_span_processor(BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint)))
    trace.set_tracer_provider(provider)
    _initialized = True


def get_tracer():
    _init_tracer_provider()
    return trace.get_tracer("iris_otel")

File: sdk/js/src/tracer.ts
import { trace, type Tracer } from "@opentelemetry/api";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";

let initialized = false;

function initTracerProvider(): void {
  if (initialized) return;

  const endpoint = process.env.IRIS_OTLP_ENDPOINT ?? "http://localhost:4318/v1/traces";
  const serviceName = process.env.IRIS_SERVICE_NAME ?? "iris-instrumented-app";

  const provider = new NodeTracerProvider({
    resource: resourceFromAttributes({ "service.name": serviceName }),
    spanProcessors: [new BatchSpanProcessor(new OTLPTraceExporter({ url: endpoint }))],
  });
  provider.register();
  initialized = true;
}

export function getTracer(): Tracer {
  initTracerProvider();
  return trace.getTracer("iris_otel");
}

/** Test-only hook, mirrors the Python SDK's escape hatch for swapping in a console exporter. */
export function _markInitialized(): void {
  initialized = true;
}
